Fix file.replace regressions, fixes saltstack#20970 and saltstack#20603

lorengordon opened the pull request:
This commit reverts to the `file.replace` code from 2014.7.0, which fixes
saltstack#20970 and saltstack#20603. It then introduces an alternative
patch that fixes saltstack#18612 and doesn't require passing through the
file twice.

There is a small regression in backup behavior... If dry run is `False`, and 
if `backup` is `False`, and if there's a pre-existing filename of `name`+`.bak`, 
then that backup file will be overwritten and deleted. This backup behavior is 
a result of how `fileinput` works. Fixing it requires either passing through the 
file twice (the first time only to search for content and set a flag), or rewriting 
`file.replace` so it doesn't use `fileinput`.

lorengordon commented:
I was coding in my sleep and came up with a better approach. Give me a bit to put it together...

lorengordon committed:
Fix file.replace regressions, fixes saltstack#20970 and saltstack#20603

This commit reverts to the file.replace code from 2014.7.0, which fixes
saltstack#20970 and saltstack#20603. It then introduces an alternative
patch that fixes saltstack#18612 and doesn't require passing through the
file twice.

There is a small regression in backup behavior... If the pattern is not
found, and if dry run is False, and if `backup` is False, and if a
pre-existing file exists with extension `.bak`, then that backup file
will be overwritten. This backup behavior is a result of how `fileinput`
works. Fixing it requires either passing through the file twice (the
first time only to search for content and set a flag), or rewriting
`file.replace` so it doesn't use `fileinput`.

lorengordon commented:
Ok, this *should* be pretty clean. It's very nearly identical to the 2014.7.0 code.

whiteinge commented:
Fwiw, pulling fileinput out completely does need to be done at some point. That module doesn't work like we need it to in a number of areas.

lorengordon commented:
Ok, several of those test files are not correct. They have blank lines, so they're actually impacted themselves by #20603.

The only test that is legitimately failing is omit_backup. It's testing whether a backup file is created even if there are no changes.

lorengordon commented:
@whiteinge, yes, I'd actually like to talk about that. The work I was doing to make file.replace hardlink friendly eliminates `fileinput`. I feel I have a pretty good handle on the logic and flow in file.replace now, and I can probably knock that out pretty quickly. The question is whether to make that change in 2014.7 (minimizing the new feature pieces like symlink and hardlink support), then pull that forward into 2015.2 and develop? Or to fix 2014.7 with something like commit #fef427b, and make the bigger changes in 2015.2 or develop? 

Either way, I'd like to finish the work on 2014.7 before getting too far ahead of this in a newer branch so I'm not stepping on my own toes with merge conflicts. :)

lorengordon commented:
Actually, it would probably make things easier if https://github.com/saltstack/salt/pull/20943 were backported into 2014.7 before trying to merge this patch in. Otherwise, pulling this one forward will require a bit of conflict resolution.

jfindlay commented:
@lorengordon, I'm wary of putting backwards-incompatible changes into 2014.7.  If it's too difficult to fix the regressions otherwise, I'm fine with the fixes going into 2015.2.

Putting the test fixes from #21091 in here, including the final end-of-lines:-), will cause the test results on this pull to turn green.

Thanks for all your work.

lorengordon commented:
I don't think any of the changes are backward incompatible, exactly...They take the same inputs and produce the same types of outputs, and just happen to fix some undesirable behavior around specific use cases.

But no matter, if we fix the file.replace regressions with this commit, then merge this forward, I'm happy to rebase PR #21030 against whatever branch you like and fix the conflicts.

I have actually been testing the file.replace code that's now in PR #21030 against the 2014.7 branch and it works great. :wink: Thank you for the dynamic module merge feature!

jfindlay commented:
@lorengordon, sure.  I'll ask @thatch45 what he thinks.

whiteinge commented:
@lorengordon pulling out `fileinput` should definitely be in 2015.2. Thanks so much for tackling that! It's been needed for a good long while. I'll defer to @jfindlay on whether it could also be in 2014.7 and then merged forward or if it should go straight into 2015.2.

lorengordon commented:
@jfindlay, I could use some help tracking down what's really going on with these jenkin test failures. I believe I was wrong about why they were failing before. After you mentioned the end-of-line issue, I realized that I needed to check how `append_if_not_found` would work with a file where the last line had a terminating LF character. As I develop on Windows, my editor shows that as a new, blank line. However, `vi` does not. So I saw that blank line at the end of the `test_replace_issue_18612_append.in` file, and thought the `append` test should append the new line after that blank line. But after a little bit of testing, that turned out not to be the case. So, the test is failing for some other reason.

I went ahead and built a virtualenv, installed salt from my branch, and ran the test manually (making a couple tweaks so it wouldn't delete the files), and as far as I can tell, the file modified by the test and the .out file in the source are exactly the same. I ran filecmp.cmp on them and it returned true. So, the test should be passing, as that's exactly what the test does. But it's failing, and I just don't know why.

lorengordon commented:
Ahh, ok, I see now. The tests are also checking whether the backup file matches the original file, and that's the part that is failing. I haven't yet addressed the backup behavior in this PR, only the regressions and the behavior described in #18612 where repeated file.replace calls with append/prepend_if_not_found were causing the file to grow indefinitely. I'll think about this a bit more and come up with something...

lorengordon committed:
Check file before making changes, create backup only if necessary

Re-add the fixes where the file is not modified if no changes are made,
and the backup is created only if changes are made.

lorengordon commented:
I think this one is good to go, finally! I'll work on getting fileinput out of 2015.2 next, then incorporate that and the hardlink friendly feature into develop. There are a couple things on my plate in the way, but it should clear up by Thursday or so this week. Most of the work is done already, and now I know how to test in my own branch in a virtualenv, so I'm hoping to knock those out pretty quick.
